compiler/src/iree/compiler/Codegen/Common/GPUTileReduction.cpp

```diff
@@ -0,0 +1,84 @@
+// Copyright 2022 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Codegen/PassDetail.h"
+#include "iree/compiler/Codegen/Passes.h"
+#include "iree/compiler/Codegen/Transforms/Transforms.h"
+#include "iree/compiler/Codegen/Utils/MarkerUtils.h"
+#include "iree/compiler/Codegen/Utils/Utils.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Linalg/Transforms/Transforms.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/Transforms/Passes.h"
+
+#define DEBUG_TYPE "iree-codegen-gpu-tile-reduction"
+
+namespace mlir {
+namespace iree_compiler {
+
+namespace {
+/// A simple pattern rewriter that implements no special logic.
+class SimpleRewriter : public PatternRewriter {
+ public:
+  SimpleRewriter(MLIRContext *context) : PatternRewriter(context) {}
+};
+}  // namespace
+
+namespace {
+
+static LogicalResult tileReduction(linalg::GenericOp op) {
+  // First split the reduction.
+  SimpleRewriter rewriter(op.getContext());
+  rewriter.setInsertionPoint(op);
+  auto control = [](linalg::LinalgOp op) {
+    linalg::SplitReductionOptions option;
+    SmallVector<int64_t> tileSize = getTileSizes(op, 1);
+    if (tileSize.empty()) return option;
+    option.ratio = tileSize.back();
+    option.innerParallel = true;
+    option.index = op.getNumLoops() - 1;
+    return option;
+  };
+  FailureOr<linalg::SplitReductionResult> result =
+      linalg::splitReduction(rewriter, op, control);
+  if (failed(result)) return failure();
+
+  unsigned numLoops = result->splitLinalgOp.getNumLoops();
+  // Then tile the new dimension to 1.
+  SmallVector<int64_t> tileSizes(numLoops, 0);
+  tileSizes[numLoops - 2] = 1;
+  linalg::LinalgTilingOptions tileOption;
+  tileOption.setTileSizes(tileSizes);
+  FailureOr<linalg::TiledLinalgOp> tiledOps =
+      linalg::tileLinalgOp(rewriter, result->splitLinalgOp, tileOption);
```

Comment:
`tileLinalgOp` is kind of being deprecated. Can we use `scf::tileUsingSCFForOp` instead?

Reply:
Thanks for catching that. Updated it.